Expose streamlines

PProfizi opened the pull request:
Create the `ansys.dpf.post.tools` package, with the `ansys.dpf.post.tools.streamlines` module.
Add the `plot_streamlines` helper to provide a first way of plotting streamlines based on a DataFrame.
Add a first test in `test_streamlines.py`.

Initial mock-up:
![image](https://github.com/ansys/pydpf-post/assets/100710998/54afa710-b8cd-49a7-9e60-1df7446aeed5)

rafacanton commented on src/ansys/dpf/post/tools/streamlines.py:
```diff
@@ -0,0 +1,76 @@
+    contour_opacity: float = 0.3,
+    **kwargs,
+):
+    """Plot streamlines based on a velocity field DataFrame.
```
```suggestion
    """Plot streamlines based on a vector field DataFrame.
```

rafacanton commented on src/ansys/dpf/post/tools/streamlines.py:
```diff
@@ -0,0 +1,76 @@
+    Parameters
+    ----------
+    dataframe:
+        A `post.DataFrame` object containing a velocity field.
```
```suggestion
        A `post.DataFrame` object containing a vector field.
```

rafacanton commented on src/ansys/dpf/post/tools/streamlines.py:
```diff
@@ -0,0 +1,76 @@
+        Expected keywords are "center", "radius", and "line_thickness".
+    mesh:
+        A `post.Mesh` object to use to compute the streamlines instead of the mesh associated with
+        the velocity field in the DataFrame.
```
```suggestion
        the vector field in the DataFrame.
```

rafacanton commented on src/ansys/dpf/post/tools/streamlines.py:
```diff
@@ -0,0 +1,76 @@
+        A `post.Mesh` object to use to compute the streamlines instead of the mesh associated with
+        the velocity field in the DataFrame.
+    plot_contour:
+        Whether to plot the field contour along with the streamlines.
```
```suggestion
        Whether to plot the field's norm contour along with the streamlines.
```

rafacanton reviewed (commented):
@PProfizi Just made a couple of suggestions to avoid referring to velocity. This is way more general and applies to all vector fields. For example, when we support EMag data, the same method could be used to plot the magnetic field lines

cbellot000 commented on src/ansys/dpf/post/tools/streamlines.py:
```diff
@@ -0,0 +1,76 @@
+    if mesh:
+        meshed_region = mesh._meshed_region
+    else:
+        meshed_region = dataframe._fc[0].meshed_region
```
we will eventually need a public method in the dataframe to get the mesh, no?

PProfizi replied:
@cbellot000 sure, but again it will need a label logic to select which field's meshed_region is of interest, or else I guess return a `MeshesContainer` as a `Meshes`

cbellot000 commented on src/ansys/dpf/post/tools/streamlines.py:
```diff
@@ -0,0 +1,76 @@
+        meshed_region = mesh._meshed_region
+    else:
+        meshed_region = dataframe._fc[0].meshed_region
+    field = dataframe._fc[0]
```
why the first field? what if you have a single time, but different label (like "elshape") in your data frame?

PProfizi commented:
@rafacanton @cbellot000 @anslpa 
Note here to keep the information:

Cannot actually create a `ansys.dpf.post.tools` sub-package with `ansys.dpf.post.tools.streamlines` module (`ansys/dpf/post/tools/streamlines.py`) as we already have a `ansys.dpf.post.tools` module (`ansys/dpf/post/tools.py`).
Three solutions:
- rename the new sub-package from `tools` to `helpers`. This is the current solution implemented, so now the new method is in `ansys.dpf.post.helpers.streamlines`. This is also what was done in `ansys-dpf-core` which makes both libraries coherent.
- Move the `streamlines.py` module directly in `ansys/dpf/post`
- Move the streamlines helpers in the `tools.py` module, which would mean no streamlines-specific module and a direct call to `from ansys.dpf.post.tools import plot_streamlines`.
- Remove the `tools.py` module and move its methods in a new module in the `tools` sub-package. This would mean makign sure that the old imports would still work. Not sure we can cover all edge-cases.

I think the first solution is the best.

What do you think?

rafacanton commented:
> @rafacanton @cbellot000 @anslpa
> 
> Just realized we cannot actually create a `ansys.dpf.post.tools` sub-package with `ansys.dpf.post.tools.streamlines` module (`ansys/dpf/post/tools/streamlines.py`) as we already have a `ansys.dpf.post.tools` module (`ansys/dpf/post/tools.py`). Four solutions:
> 
> * rename the new sub-package from `tools` to `helpers`. This is the current solution implemented, so now the new method is in `ansys.dpf.post.helpers.streamlines`. This is also what was done in `ansys-dpf-core` which makes both libraries coherent.
> * Move the `streamlines.py` module directly in `ansys/dpf/post`
> * Move the streamlines helpers in the `tools.py` module, which would mean no streamlines-specific module and a direct call to `from ansys.dpf.post.tools import plot_streamlines`.
> * Remove the `tools.py` module and move its methods in a new module in the `tools` sub-package. This would mean makign sure that the old imports would still work. Not sure we can cover all edge-cases.
> 
> I think the first solution is the best.
> 
> What do you think?

@PProfizi I agree, I think it is right to have both libraries with a similar structure. Thanks!

PProfizi commented:
> > @rafacanton @cbellot000 @anslpa
> > Just realized we cannot actually create a `ansys.dpf.post.tools` sub-package with `ansys.dpf.post.tools.streamlines` module (`ansys/dpf/post/tools/streamlines.py`) as we already have a `ansys.dpf.post.tools` module (`ansys/dpf/post/tools.py`). Four solutions:
> > 
> > * rename the new sub-package from `tools` to `helpers`. This is the current solution implemented, so now the new method is in `ansys.dpf.post.helpers.streamlines`. This is also what was done in `ansys-dpf-core` which makes both libraries coherent.
> > * Move the `streamlines.py` module directly in `ansys/dpf/post`
> > * Move the streamlines helpers in the `tools.py` module, which would mean no streamlines-specific module and a direct call to `from ansys.dpf.post.tools import plot_streamlines`.
> > * Remove the `tools.py` module and move its methods in a new module in the `tools` sub-package. This would mean makign sure that the old imports would still work. Not sure we can cover all edge-cases.
> > 
> > I think the first solution is the best.
> > What do you think?
> 
> @PProfizi I agree, I think it is right to have both libraries with a similar structure. Thanks!

What I ended-up doing:
- add a new `ansys.dpf.post.helpers` sub-package
- add a new `ansys.dpf.post.helpers.streamlines` module
- add a new `ansys.dpf.post.helpers.selections` module
- moved the `ansys.dpf.post.tools.select` method to the `ansys.dpf.post.helpers.selections` module
- made an import to it in `ansys.dpf.post.tools` to ensure retro-compatibility
- updated the dosctring in `ansys.dpf.post.tools` to mention the move to `ansys.dpf.post.helpers`

cbellot000 commented on src/ansys/dpf/post/helpers/__init__.py:
```diff
@@ -0,0 +1,6 @@
+"""Tools package.
+
+Tools
```
Helpers?

cbellot000 replied:
Is there a plan for follow up PR adding an example for streamlines?

PProfizi replied:
> The only problem I see is naming. For me "helpers" means wrapping already existing capabilities in another way to make them easier to be used

IMO this is what this method is? The end-goal is to have a much more thorough streamlines manipulation capability, this method being a quick way to plot the streamlines given a Dataframe.



> Is there a plan for follow up PR adding an example for streamlines?

Yes of course, I initially wanted this merged before last Friday's deadline, but if the deadline is not an issue anymore, I can add and example within this PR.

PProfizi replied:
> The only problem I see is naming. For me "helpers" means wrapping already existing capabilities in another way to make them easier to be used

@cbellot000 What I can do though is rename `helpers` as `tools`, but that will require deleting the `tools.py` module, which I think can be done without retro-compatibility issues if I import the `select` method from within the `__init__.py` file of the new `tools` sub-package, yet I may have missed a case where it breaks.

cbellot000 replied:
@PProfizi no, it's okay, your justification makes sense
